### ISSUE TITLE
GAMS is not registered correctly in show-versions

### ISSUE THREAD

**Issue by glatterf42** [iiasa/ixmp#523]
#### Code sample or context

Recently, @behnam-zakeri tried to install message-ix again on a new Linux machine. The process went fine by creating a new virtual environment via virtualenv and virtualenvwrapper and we made sure that the directory for `gams` is in `$PATH`. This can be confirmed by running `gams` on the system; it seems to work both from the command line and as part of the westeros_baseline tutorial from message-ix. However, running 

```bash

message-ix show-versions

```
tells us that `'gams' is not in PATH`. Since everything else is working fine, my expectation is that this is due to an issue with the way that ixmp tries to find `gams`, not about the installation per se.

We should note that the Linux machine was not a fresh install. Things like `pip` and `virtualenv` or even `venv` were missing, which seems odd. 

#### Expected result

`message-ix show-versions`, which is internally using https://github.com/iiasa/ixmp/blob/d202e019204e6cf979df6f2a827322a00dbd1957/ixmp/model/gams.py#L18C1-L50C20 should find the correct version of `gams`. 

#### Problem description

It doesn't seem to find it at all.

#### Versions

@behnam-zakeri If you can edit this description, please add the output of `message-ix show-versions` here. Otherwise, please just paste it down below.

<details><summary>Output of <tt>ixmp show-versions</tt></summary>

```
<!--
Run one of the following and paste the results here:
- 'ixmp show-versions' in a terminal, or
- 'import ixmp; ixmp.show_versions()' in a Python interpreter.
-->
```

</details>

**Comment by khaeru** [comment]
One way to debug this is to put for instance:
```python
    print(os.environ)

    # Execute, capturing stdout
    output = check_output(
        ...
    )

    print(output)
```
…around the subject code.

Setting `PATH` can have some subtleties. For instance, if setting the value in .profile or a similar file, the value will appear in all new terminals opened _after_ that addition. But if one has a terminal open prior, and does not do anything in it, then its prior value of `PATH` is not automatically updated.

**Comment by glatterf42** [comment]
Hi @Wegatriespython, thanks for picking this up :)
As you can see, our discussion above stalled, so we don't yet have more information about what's going on and how we could fix this issue. 
I see that you installed message-ix-models and message_data from source, so you know how this works. Could I ask you to please also install ixmp and message_ix from source? For this, you might need to start from a fresh virtual environment to make sure no traces of either are left when you start this process. This would allow you to implement what @khaeru [suggested above](https://github.com/iiasa/ixmp/issues/523#issuecomment-2014503003) and report back the output, which would then (hopefully) allow us to get to the bottom of this issue :)

**Comment by glatterf42** [comment]
One more clarifying question: why did you restart VSCode multiple times? You said this error occurs in your PowerShell session, so if you need to restart anything to make sure the changes to your `PATH` take effect, it would be your PowerShell session, not VSCode.

**Comment by glatterf42** [comment]
Lastly, if I remember correctly, this discussion stalled because even though `message-ix show-versions` didn't pick up GAMS correctly, the rest of the code did. So could you please try running one of the tutorials to see if you can actually solve a `Scenario`? 
If you can, you might not need any workaround after all :)

**Comment by Wegatriespython** [comment]
I will look into this and return. Ah I was using the integrated terminal in VScode (set to powershell) so I did the restart to ensure the powershell started with a fresh session (tried setting persistent terminal sessions to off).

**Comment by Wegatriespython** [comment]
Fixed the issue will open PR soon. 

Issue:
The original code used a list of arguments with shell=True and relied solely on the command name "gams". On Windows, this can fail if the command must be resolved in an environment where the working directory isn’t on the same drive as the executable.

Fix:
By using shutil.which, we obtain the absolute path to the GAMS executable and then build a proper string command. This ensures that the executable is correctly located and called regardless of the working directory’s location.

Issue : 
```python 
            try:
                # Execute this no-op file and capture stdout
                output = check_output(
                    ["gams", "null.gms", "-LogOption=3"],
                    shell=os.name == "nt",
                    cwd=temp_dir,
                    universal_newlines=True,
                )
```

Fix  : 
```python 
            try:
                gams_exec = os.getenv("GAMS_EXECUTABLE", "gams")
                if os.name == "nt":
                    import shutil  # get full path on windows
                    found = shutil.which(gams_exec)
                    if found:
                        gams_exec = found
                    else:
                        gams_exec = "gams"
                    cmd = f'{gams_exec} null.gms -LogOption=3'
                else:
                    cmd = [gams_exec, "null.gms", "-LogOption=3"]
                output = check_output(
                    cmd,
                    shell=os.name == "nt",
                    cwd=temp_dir,
                    universal_newlines=True,
                )
```

**Comment by khaeru** [comment]
@Wegatriespython —good catch!

If `shutil.which()` is cross-platform, do we even need to guard it with `if os.name == "nt"`? This seems to lead to unnecessary duplication.

**Comment by Wegatriespython** [comment]
If I understand correctly we need the special case for windows, since windows expects a string while linux and osx can take a list. Then at check_output we need the os check to set the shell true or false. 

```python 
gams_exec = os.getenv("GAMS_EXECUTABLE", "gams")
# get full path if available 
gams_exec = shutil.which(gams_exec) or gams_exec

# build string if windows 
if os.name == "nt":
    # Windows requires a single string command when shell=True.
    cmd = f'{gams_exec} null.gms -LogOption=3'
else:
    # POSIX systems can safely use a list of arguments.
    cmd = [gams_exec, "null.gms", "-LogOption=3"]

output = check_output(
    cmd,
    shell=(os.name == "nt"),
    cwd=temp_dir,
    universal_newlines=True,
)
```

**Comment by khaeru** [comment]
> I understand correctly we need the special case for windows,

Can you please mention where you got that info?

It seems to contradict the [docs for `subprocess.Popen`](https://docs.python.org/3/library/subprocess.html#subprocess.Popen), which say:

>  Unless otherwise stated, it is recommended to pass args as a sequence.

> On Windows, if args is a sequence, it **will be converted** to a string in a manner described in [Converting an argument sequence to a string on Windows](https://docs.python.org/3/library/subprocess.html#converting-argument-sequence). This is because the underlying CreateProcess() operates on strings.

**Comment by Wegatriespython** [comment]
Makes sense didn't know the underlying implementation for subprocess. Since it automatically does the conversion when a sequence is provided, it should be fine to supply the sequence in both cases. I passed the windows as a string just to eliminate any confounders while debugging but ig we can remove the extra check.
